### ISSUE TITLE
Automatically disable dormant users.

### ISSUE THREAD

**Issue by Scuilion** [keycloak/keycloak#11800]
### Description

The feature requested is to automatically disable a users account after a configurable period of inactivity. This is a common security practice in enterprises to guard against a dormant users credentials being discovered and used without approval.

By inactivity, we mean that the user has not logged in directly to keycloak or has not obtained access tokens via OpenIDConnect.

The period of inactivity should be a configurable setting and would typically be set to a number of days (e.g. 60, 120, 180 days.). Any activity by the user would reset his inactivity timer. The setting may be at realm level. If no value is set, then the feature is off.

If the inactivity period has elapsed without any activity by a user, then the users "Enabled" setting would be automatically set to "off". The account can only be re-activated by the administrator.

### Discussion

https://issues.redhat.com/browse/KEYCLOAK-5865

**Comment by Scuilion** [comment]
Can the EVENT_ENTITY table be used to accomplish this task in the background?

**Comment by SomeDeveloper13** [comment]
Since the length of time those records are kept is configurable and disablable. Probably not a good idea. What's more, the logging method would frequently need to be customized. Which would mean that the table isn't necessarily used at all in such a case.
Maybe consider a new option in the authentication flows?

**Comment by anuragpathak21** [comment]
Last login time should be kept with the user profile and can be used to perform this action. It's  a common requirement that alomost every customer ask for. 
Currently achieve it, we have added event listener spi and keep the last login informtaion on application side, we then have another process which checks this value and disable the users who have not logged insince n days, where n is configured on client account.
It would be great to have this inbuild.

**Comment by icc-garciaju** [comment]
For my use case it would be great if:
- Accounts can be requiered to re-verify their emails
- Unverified accounts are removed after a per realm configurable amount of time

**Comment by larmbuster** [comment]
Has there been any work on this? It would an amazing capability. I've also looked into automating this behavior through Ansible but it doesn't look like the API supports checking last login timestamps.

**Comment by greghall76** [comment]
This is a common security compliance kind of requirement. Very much hoping for progress.

**Comment by JohnGlacken1** [comment]
Is this feature available now?  If not, is there a timeline when it will be released (if at all)?

**Comment by gambled** [comment]
I am also interested on the status of this potential feature.  This is/has become a common control in most of the security frameworks (SOC-2, Hi-Trust, FedRamp, etc).  Would rather use out-of-the-box functionality than roll our own.

**Comment by sschu** [comment]
I guess the Keycloak team will happily accept contributions.

**Comment by vitorcont** [comment]
Are there any updates on this? Very much hoping for this feature

**Comment by ataraxus** [comment]
Similar requirement from BSI Grundschutz ORP.4.A5 Vergabe von Zutrittsberechtigungen (german security framework). Just sayin, this is quite common and sent me on this path ;)

**Comment by andreconrado** [comment]
Another one here who will love this feature

**Comment by nunosimoes-gls** [comment]
Feature much needed, one less workaround needed.

**Comment by anuragpathak21** [comment]
Hey team, any plan to implement this?

**Comment by stianst** [comment]
We don't have any immediate plans to support this I'm afraid, but if someone is interested in contributing something let's talk.

**Comment by col-panic** [comment]
What do you think about the following solution:

* (ADMIN UI) Add a new Tab `User Activity Policy` under `Authentication/Policies` with the following settings:
  * _Disable inactive users_ (boolean) [Disable user with no login activity after a certain time period]
  * _Disable after_ (integer) [The number of days to disable the user after]
  * _Inform user_ (boolean) [Send a mail to the disabled user]
* (SERVER) Implement a respective SPI in the backend that performs a once-a-day async cronjob to check for each user when the user last successfully logged in via `EVENT_ENTITY` table entries for the user of type `LOGIN` (with all the rest endpoints etc) and creates a resp. DISABLE BY POLICY event by itself. (An SQL index  on EVEN_ENTITY#TYPE and EVENT_ENTITY#USER_ID might be helpful) 

This is just a naive approach considering LOGIN event. @stianst as you remarked, I would also not automatically delete the user, maybe as a configurable option with an even later time do to so. There is no elaborate caring on open offline sessions and so on - but just a bare prevent of a direct user login after the time by disabling the user.

For our current use case such a solution would fit.

**Comment by luizhenrique-dev** [comment]
This feature is really important for my use case. I hope the Keycloak team or the community can implement it soon.

**Comment by stianst** [comment]
> What do you think about the following solution:
> 
> * (ADMIN UI) Add a new Tab `User Activity Policy` under `Authentication/Policies` with the following settings:
>   
>   * _Disable inactive users_ (boolean) [Disable user with no login activity after a certain time period]
>   * _Disable after_ (integer) [The number of days to disable the user after]
>   * _Inform user_ (boolean) [Send a mail to the disabled user]
> * (SERVER) Implement a respective SPI in the backend that performs a once-a-day async cronjob to check for each user when the user last successfully logged in via `EVENT_ENTITY` table entries for the user of type `LOGIN` (with all the rest endpoints etc) and creates a resp. DISABLE BY POLICY event by itself. (An SQL index  on EVEN_ENTITY#TYPE and EVENT_ENTITY#USER_ID might be helpful)
> 
> This is just a naive approach considering LOGIN event. @stianst as you remarked, I would also not automatically delete the user, maybe as a configurable option with an even later time do to so. There is no elaborate caring on open offline sessions and so on - but just a bare prevent of a direct user login after the time by disabling the user.
> 
> For our current use case such a solution would fit.

I don't think that would scale very well considering how events are stored in Keycloak currently. It would also require keeping event history for at least as long as the grace period. I would imagine you'd want to configure users to be marked as dormant after many months, maybe a year; which would require storing all events for that duration.

With millions of users and events, the cronjob could also be very slow unless there is proper indexes in place, which is a bit problematic to add, because adding indexes on existing tables with millions of entries is slow, and would be problematic from an upgrade perspective to many; for this we have considered in the past figuring out some way of doing some of the migration in the background or on a schedule, but we haven't had a chance to implement that yet.

**Comment by stianst** [comment]
We do plan to pick this up at some point, but it may be a while until we do. I do believe this would be a bit of a fair amount of effort to implement properly, but if someone from the community are interested, let me know and we can try to come up with a design around it that would work.

**Comment by CyberPoly1000** [comment]
Would be a very nice feature due to compliance requirements!

**Comment by gndctl-matt** [comment]
+1 for this feature - required for compliance. Looks like I will have to implement myself externally for now

**Comment by michelkluger** [comment]
big +1 for this feature

**Comment by Ricq** [comment]
Yes, please! +1.

**Comment by sharifmamun** [comment]
Yes, please. This is critical for security compliance and thousands of new organizations are eagerly waiting for this feature in `Keycloak`.

**Comment by stianst** [comment]
We are working on support for automatic disable/removal of users. There is probably a duplicated issue for it. @pedroigor @sguilhen @vramik could you provide the link and perhaps close this one as a duplicate or something?

**Comment by stianst** [comment]
See https://github.com/keycloak/keycloak/issues/39888
